### PR TITLE
Go bindings: CI gate, review test gaps, and remaining low-severity findings

### LINKED ISSUE #274: Go bindings are invisible to CI and have untested failure paths
No workflow runs the Go bindings at all. Module tests, the 32-bit builds that the u32 bounds guards exist to protect, and the examples module (a separate Go module, so it falls outside every `go test ./...` sweep) are all invisible to CI — which is how a broken documented read path was able to ship green. Every Go surface needs a CI job, including on stacked PRs.

A CI job should vet and test both Go modules with `CGO_ENABLED=0`, cross-compile the value package for a 32-bit target, and run the documented example end to end against its database.

Alongside the gate, the error and rejection branches flagged during review need tests rather than assumed behaviour: duplicate-key rejection in the ciphertext decoder (reachable only from raw bytes), malformed ciphertext inputs (depth bombs, hostile counts, trailing bytes, unknown tags), encode-side depth bombs and unsupported kinds, non-string map keys, the value-conversion and container error arms, and the remaining scan cases — plus a white-box pin on the eager-capacity clamp mirroring the Rust-side fix.

Two remaining low-severity findings should be closed in the same pass: an encoder that commits a map key before its value should fail at encode time naming the offending key instead of shifting the wire and surfacing later as an opaque malformed-input error, and the static absence-verification path should require the sealed plaintext to be empty so a key-holding writer cannot pass a nonempty payload off as an absence marker.

Addressed by #264 (stacked on #257).

### PR DESCRIPTION
**Pull request by coderdan**
Stacked on cipherstash/vitaminc#257. Three strands, one per commit group:

## CI (\`go.yml\`)

No workflow ran \`bindings/go\` at all — module tests, the 32-bit builds the u32 guards protect, and the examples module (its own Go module, outside every \`go test ./...\` sweep) were all invisible to CI, which is how cipherstash/vitaminc#257's broken documented read path shipped green. The new job vets and tests both modules (\`CGO_ENABLED=0\`), cross-compiles \`vcvalue\` for \`GOOS=linux GOARCH=386\`, and runs the \`userrecord\` example end to end.

## Review test gaps (from cipherstash/vitaminc#257's test-coverage review)

All eight flagged branches now have tests: duplicate-key rejection in \`decodeCipherText\` (raw-bytes, unreachable from Go maps), \`UnmarshalCipherText\` malformed inputs (depth bomb / hostile count / trailing bytes / unknown tag), \`Unmarshal\` array+object depth bombs and hostile counts, encode-side depth bomb, unsupported kinds, non-string map keys, \`Plain.Value\` conversions incl. the uint64-overflow and container errors, and the remaining \`Sealed.Scan\` arms. Plus a white-box \`eagerCap\` pin mirroring the Rust \`eager_capacity\` mutant fix.

## Remaining low-severity findings

* **\`MapEncoder.Field\` key-before-value commit**: a skipped value now fails at encode time with the entry's key named, instead of shifting the wire and surfacing as an opaque \`errMalformed\` at decode. Same check for \`SeqEncoder.Elem\`. The two dispatchers' marker interception is also folded into one \`interceptMarkers\` helper.
* **\`verify_absent\` tag-only check**: the static path now requires the sealed plaintext to be empty, so a key-holding writer cannot pass a nonempty payload off as an absence marker.
* **No change, with rationale**: \`Aad::pae\`'s \`pub\` visibility is required by \`vitaminc-aead-value\`'s leaf-type derivation and already documented as the upstream building block; the \`seal_leaf\`/\`encrypt_bytes_array\` duplication is a deliberate no-realloc fast path, documented in place and pinned byte-equal by test.

Closes cipherstash/vitaminc#274

### CODE REVIEW

File: bindings/go/vcvalue/encoder.go [hunk 1]
```diff
@@ -274,19 +310,27 @@ type MapEncoder struct {
...
 }
 
 // Field writes an entry key and returns an Encoder for its value. Write the
 // value fully (one channel) before calling Field again. Keys are UTF-8
 // validated; insertion order is preserved on the wire.
 func (m *MapEncoder) Field(key string) Encoder {
 	if m.st.err == nil {
-		if !utf8.ValidString(key) {
+		if m.count > 0 && m.st.nodes == m.mark {
+			m.st.fail(fmt.Errorf("vcvalue: value for the entry before %q was never written", key))
```
[the remark this answers is not in the document]
Comment (coderdan) [reply bindings/go/vcvalue/encoder.go]: Fixed in 666b55a: `MapEncoder` now tracks `lastKey` and the diagnostic names the key whose value slot misbehaved (`value for key "a" was never completed` / `more than one value was written for key "a"`). Sequence diagnostics carry the element index. Message assertions added to `TestEncoderCatchesSkippedValuesAtEncodeTime`.

File: bindings/go/vcvalue/encoder.go [hunk 2]
```diff
@@ -44,6 +44,13 @@ type Plain struct {
 type encState struct {
 	buf []byte
 	err error
+	// nodes counts completed value nodes (terminals, and containers at their
+	// End). MapEncoder.Field and SeqEncoder.Elem record it to verify their
+	// "write exactly one value before the next entry" contract at encode
+	// time — without this, a skipped value shifts the wire (the next key
+	// parses as the missing value) and only surfaces as an opaque
```
[the remark this answers is not in the document]
Comment (coderdan) [reply bindings/go/vcvalue/encoder.go]: Fixed in 666b55a. The global node counter is gone: each `Encoder` now carries a pointer to its slot's completion counter, and containers require exactly one completed **direct** child per `Elem`/`Field` (verified again at `End`). This catches un-Ended nested containers, doubly-written slots, and — via the same rule applied to the root slot in `Marshal` and the passthrough payload path — un-Ended root containers and zero/multiple root writes. `Passthrough` and pointer/interface dereference deliberately share the parent slot's counter, since they add no value node of their own. New table cases cover all the scenarios you listed.

File: bindings/go/vcencrypt/encoder.go [hunk 3]
```diff
@@ -250,21 +278,51 @@ type SeqEncoder struct {
...
+	done *int
+	// childDone counts completions of this sequence's own element slots.
+	// Elem and End verify childDone == count, so a skipped element, an
+	// un-Ended nested container, or a doubly-written element fails here at
+	// encode time instead of as errMalformed at decode.
+	childDone int
+}
+
+// checkPrev verifies the previous element slot completed exactly once.
+func (s *SeqEncoder) checkPrev() {
```
Comment (freshtonic) [inline bindings/go/vcencrypt/encoder.go:281]: Per-container `childDone` is the right fix here. Checking both `< count` and `> count` catches skipped values, un-ended nested containers, and multiple writes to the same slot—cases a global completed-node counter cannot distinguish.

**Review by freshtonic — APPROVED** [review]
## Standards

Pass. The workflow covers both Go modules, the separate example module, `CGO_ENABLED=0`, and the 32-bit target. The encoder fix replaces the previously global completion signal with per-slot state rather than layering on another heuristic.

## Spec

Pass. The listed malformed-input/test gaps, eager-capacity pin, named skipped-value errors, exact-one slot completion, and empty absence-marker verification are all present. This also closes the only qualification noted on #257.

Approved. CI is green and I found no remaining blocking issue in the stack.

File: bindings/go/vcencrypt/ciphertext.go [hunk 4]
```diff
@@ -123,10 +123,14 @@ func encodeCipherText(out []byte, v any, depth int) ([]byte, error) {
 		// recursion budget with the surrounding ciphertext (depth+1),
 		// mirroring the Rust CT_PASSTHROUGH framing.
 		st := &encState{buf: out}
-		encodeAny(Encoder{st: st, depth: depth + 1}, n.V)
+		var done int
+		encodeAny(Encoder{st: st, depth: depth + 1, done: &done}, n.V)
 		if st.err != nil {
 			return nil, st.err
 		}
+		if done != 1 {
```
[the remark this answers is not in the document]
Comment (coderdan) [reply bindings/go/vcencrypt/ciphertext.go:133]: Added in cf8fe73: `TestMarshalCipherTextRejectsPayloadMiscount` drives all three malformed payloads (zero writes, two writes, un-Ended container) wrapped in `Plain` through `MarshalCipherText` — verified the branch is live (test fails with the check removed).

File: bindings/go/vcencrypt/encoder.go [hunk 5]
```diff
@@ -207,12 +233,14 @@ func (e Encoder) Passthrough() Encoder {
 			e.push(tagPassthrough)
 		}
 	}
-	return Encoder{st: e.st, depth: e.depth + 1}
+	// The passthrough marker and its wrapped value form ONE value in this
+	// Encoder's slot, so the wrapped value completes the same counter.
+	return Encoder{st: e.st, depth: e.depth + 1, done: e.done}
```
[the remark this answers is not in the document]
Comment (coderdan) [reply bindings/go/vcencrypt/encoder.go:227]: Added in cf8fe73: `TestEncoderPassthroughRequiresWrappedValue` writes a `Passthrough()` marker into a sequence slot with no wrapped value and requires an encode-time failure naming `element 0`.

**Review by auxesis — APPROVED** [review]
Solid follow up PR to #257, thanks @coderdan — approved.

Four findings from that review touch files this PR doesn't (`vcvalue/sealed.go`, `vcvalue/decode.go`, `vcencrypt/client.go`), so they're staying open on #257 rather than being fixed here. Flagging them here too so they don't get lost once #257 merges:

- **`Plain.Value()` missing narrow int types** — [comment](https://github.com/cipherstash/vitaminc/pull/257#discussion_r3781742598). The type switch handles `int32`/`uint32`/`int64`/`uint64`/`float32`/`float64` but not `int`/`int8`/`int16`/`uint8`/`uint16`, so `vcvalue.Plain{V: user.ID}` with a plain `int` field fails at the driver boundary even though the encoder handles `int` fine.
- **`tagObject` decode has no duplicate-key check** — [comment](https://github.com/cipherstash/vitaminc/pull/257#discussion_r3781742607). `decodeCipherText`'s `ctMap` case rejects duplicate keys; `decodeValue`'s `tagObject` case (the plaintext decoder) doesn't. A decoded `Object` can carry duplicate keys that silently collapse on re-encode.
- **`allocWrite` leaks the guest buffer on write failure** — [comment](https://github.com/cipherstash/vitaminc/pull/257#discussion_r3781742615). If `Memory().Write` fails after `vc_alloc` already succeeded, the zero-value `guestBuf{}` is returned instead of the real one, so the caller's deferred `free` no-ops and the allocation is never released.
- **No `WithCloseOnContextDone`** — [comment](https://github.com/cipherstash/vitaminc/pull/257#discussion_r3781742622). A caller's `ctx` deadline/cancellation doesn't interrupt an in-flight guest call, and `Client` serializes all calls behind one mutex, so one slow call blocks every other caller with no way out via `context`.

None of these are blocking for this PR — I just don't want them to fall off the radar.

**Comment by coderdan** [comment]
Two additions from the PR #257 security review, landed here because this PR owns the encoder rewrite and the CI workflow (details in the [summary on #257](https://github.com/cipherstash/vitaminc/pull/257#issuecomment-5300759095)):

- **`8089292`** — `MapEncoder.Field` rejects duplicate keys at encode time. `encodeStruct` could emit duplicate wire keys (a `vc` tag colliding with another field's effective name) that only failed after the plaintext crossed the FFI boundary, as an opaque `errMalformed`. The check sits in `Field` so the struct, map, and hand-written `Encryptable` routes are all covered, with a diagnostic naming the colliding key.
- **`98b52fb`** — the 386 check now *runs* the vcvalue tests instead of only building (a build check cannot catch the negative-count `make()` panic class the guards exist for), and a new `guest` job gates the wasm guest crate — fmt, clippy on both the wasm32-wasip1 ABI build and the native test build, and the native unit tests. That crate holds all the unsafe FFI code and previously had zero CI.

The branch also merges the updated `spike/go-wasi` (the guest ABI hardening + duplicate-key codec fixes), resolving the test-file overlap: the two eagerCap pins merged into one table test, and both Plain.Value test angles kept.

— Claude Fable (Anthropic's Claude Fable 5 model) via Claude Code, at Dan's request.

**Comment by coderdan** [comment]
@auxesis closing the loop on the four carry-overs you flagged — all four were fixed in 22d124b on #257 shortly after your approval, and that commit is now merged into this branch as well, so nothing is left open on either side:

- **`Plain.Value()` narrow int types** — every integer kind the encoder accepts now converts (`int`/`int8`/`int16`/`uint8`/`uint16` added, `uint` with an overflow check), with a table test over all arms.
- **`tagObject` duplicate-key check** — the plaintext decoder now rejects duplicate wire keys exactly as the `ctMap` arm does, with a test; the Rust codec gained the matching check in 6a212c4 so the two sides agree.
- **`allocWrite` leak** — the guest allocation is freed before returning the write-failure error.
- **`WithCloseOnContextDone`** — set in the runtime config, with `TestContextCancellationFailsGuestCall` covering the interrupt path.

— Claude Fable (Anthropic's Claude Fable 5 model) via Claude Code, at Dan's request.
